### ISSUE TITLE
What is the correct way to update another environment from Feature Branch?

### ISSUE THREAD

**Issue by danrivett** [aws-amplify/amplify-cli#3898]
**Which Category is your question related to?**
Deployments with Feature Branches. 

I may be doing something wrong, but after reading the docs, I'm still not clear if I am doing anything wrong or have wrong expectations possibly because the docs aren't clear enough?

So I've created this question to verify this.

**Amplify CLI Version** 
4.18.0

**What are you attempting to do?**
- We have a multi-developer React Native Amplify Project setup.
- We have 2 backend environments initially deployed from `master` which just has Auth added.
- Developer A created a new feature branch and added an AppSync API Category to the backend
  - They pushed it to their own backend environment successfully.
  - After testing they created a Pull Request for review
- Developer B attempted to checkout the feature branch containing the AppSync API and update their own backend environment to deploy the new API contained in the Feature Branch
  - Developer B's backend environment was last deployed from `master` and so didn't previously have the API component.
- However after doing a `git checkout <feature-branch>` followed by an `amplify status`, it showed no new API to push
  - Therefore `amplify push` did nothing and didn't deploy the new API to their backend.

**Current Resolution**
The only way `amplify status` would show the new API checked in under the Feature Branch after checking out the Feature Branch was to do an `amplify init` from that Feature Branch even though `amplify init` was previously run earlier whilst on the `master` branch (which doesn't contain the new API configuration).

**Expected Behavior**
From reading the docs, especially the [Environments and Teams](https://aws-amplify.github.io/docs/cli-toolchain/quickstart#environments-and-teams) section of the docs, I expected to be able to simply checkout the feature branch and do an `amplify status` and see the changes checked into the feature branch but not in the current environment.

See below for more detail:

**Developer A**

```
git clone ...
amplify init # Create new developerAEnvironment
amplify push # Push new developerAEnvironment to create new backend environment
# Currently backend just has an Auth category deployed

git checkout -b dev-a-branch
amplify add api ...
amplify push
# Now this backend environment also has an API category deployed

git add -A
git commit -m "Look at my new GraphQL API!"
git push -u ...
```

So far so good up to this point...

**Developer B**

```
git clone ...
amplify init # Create new developerBEnvironment
amplify push # Push new developerBEnvironment to create new backend environment
# Currently backend just has an Auth category deployed

git checkout dev-a-branch
amplify status
# Expect to see new API category to deploy from the API config checked into dev-a-branch
# Don't see any API category to deploy

amplify init # Again. Once again select the existing developerBEnvironment
amplify status
# Now see the new API category how up in 'Create' status, so can now push

amplify push
# New API category gets deployed to developerBEnvironment as expected
# Just didn't expect to have to do the 'amplify init' a second time above
```

**Comment by SwaySway** [comment]
Hello @danrivett 
Based on this workflow it seems each developer creates their own sandbox environment to test changes before committing them to a dev branch? If that is the case then we do have documentation on sandbox environments. https://docs.amplify.aws/cli/teams/sandbox

**Comment by danrivett** [comment]
Thanks @SwaySway for the info. Our scenario does look like the "Sandbox" workflow but with one key difference which **I think uncovers a bug in Amplify here**:

The Sandbox workflow assumes the same developer added the additional Amplify category to the sandbox branch before switching to the other branch and new Amplify environment, and merging the code.

If the same developer does this, it works - I just validated that. 

It works because some Amplify config file that's not under source control (I'm not sure which files) knows about the new Amplify category from before both the other branch and backend environment which don't have it is checked out.

With that flow once the 'sandbox' branch is merged in, as it already knew about the new Amplify Category, it shows up as 'Create' for the new backend environment checked out.

---

**However, this does _not_ work if there are two different developers doing this**, since that second developer never has had a backend environment deployed with the new Amplify category, and so doesn't have the transient Amplify config not under source control.

Therefore even though there is a new Amplify category in the feature branch that has either been checked out (or merged into `dev` like in the example), `amplify status` won't show the new Category.

I ask you to please recreate this yourself, I am 99% sure this a bug in Amplify based on extensive  testing I've done on this over several hours, and you should be able to create it very easily if you follow the steps clearly.

The key to recreate is to simulate two different developers by having a second clone of the repostory locally so you don't have transient Amplify config from the previous feature-branch Amplify environment which contained the new Category.

Once you do that even though you'll have checked out the feature/sandbox branch with the new Amplify category checked in under source control, you won't see the new Amplify Category show up unless you do an `amplify init` to regenerate the transient files from the feature branch.

It's been a long day investigating this bug so apologies this doesn't make much sense, but hopefully it makes enough sense for you to recreate the problem.

**Comment by kaustavghosh06** [comment]
@danrivett Have you looked into our `amplify pull` command which we introduced recently? I suggest your team take a look at that - https://docs.amplify.aws/cli/teams/multi-frontend . I believe this should help with the issue that you're describing above

**Comment by danrivett** [comment]
@kaustavghosh06 Yes I have. It doesn't help in this scenario since the new Amplify Category being added is contained in the feature branch and we want to deploy it to our environment, not the other way round, so there's nothing to pull so to speak.

We want to make sure we only deploy in a configuration-as-code manner so being able to retrieve new definitions from source control and deploy them into environments should work.

The fact that doing an `amplify init` after checking out the source control seems to indicate to there's some transient files that amplify has a dependency on, it just seems odd to do a second `init` command and may lead to unintended side effects if that's not the intention of the Amplify team.

I think the best thing is to recreate this issue as it makes more sense why I created this issue once you've recreated it, before then it may seem odd.

**Comment by kaustavghosh06** [comment]
@danrivett Okay, so you haven't pushed the new category - but want the other developers to have access to it? 
Why not share the files via git and then run `amplify init` -> `select the env`
This would show up as "Create" for the new resource added until you push it to the cloud

**Comment by danrivett** [comment]
So to clarify that workflow you linked to @kaustavghosh06 is for multiple front-ends, we actually have the opposite here, multiple back-ends which is why it doesn't work.

What I'm describing can be summed up as the git Pull Request workflow:

- The new Category has been pushed by Developer-A, but to developer-A's own environment during their implementation.
- Developer-B has their own environment and has checked out the feature branch with the new Category defined and wants to push that new config to their own environment.
  - It could also be the CI system that responds to the Feature Branch being merged into master, the same would apply.

We don't want to pull runtime config from Dev-A's backend environment to push it into Dev-B's environment because we want it to be managed via source control, Dev-A's backend could contain other changes they've made not related to this feature branch.

This workflow does work with the additional step of doing an 'amplify init' the second time once the feature branch has been checked out, as the transient files on Dev-B's machine now contains new files based purely on the extra configuration added in the Feature Branch.

If this is the expected solution we can use it. It's just not documented that way currently so I want to make sure we don't run into side effects or this gets broken in the future if it's not intended to be used in this manner.

If it is the intended approach I would suggest the documentation is updated to describe the scenario since both the [sandbox](https://docs.amplify.aws/cli/teams/sandbox) workflow and the   [team workflow](https://aws-amplify.github.io/docs/cli-toolchain/quickstart#environments-and-teams) only work if the same developer does the `amplify push` (or if they've pulled from the other environment they've never worked on).

I know this may seem odd, but the infrastructure-as-code aspect I feel is very important to avoid unexpected config changes percolating between systems as developers play with the environments, especially their own ones, and so can introduce other temporary changes never intended to be migrated and not included in any Pull Request.

Shockingly they may be working on two User Stories at the same time so their environment may contain other unrelated changes not ready for inclusion in the PR.

**Comment by danrivett** [comment]
> Why not share the files via git and then run `amplify init` -> `select the env`
> This would show up as "Create" for the new resource added until you push it to the cloud

Your last comment as I understand it is exactly what we're doing, it's just not documented that this is intended. As I've understood the documentation, `amplify init` is only described as to be used on a fresh uninitialized project such as after a `git clone` etc.

**Comment by kaustavghosh06** [comment]
Yeah, we can be expicit about this in our use-case - stating that you should execute `amplify env checkout <env-name>` (which is the same as amplify init -> select env) after a git merge for merging categories which aren't pushed to the cloud.

**Comment by danrivett** [comment]
Thanks @kaustavghosh06.

I think our scenario (which I think isn't a crauzy uncommon one) is that there is already an active environment and we've just received new config either from checking out a new branch, or in a CI environment, from having it merged into master.

In a CI environment things are starting from scratch and so an `amplify init` will likely always be done, so no problem there.

In a local developer environment, they have likely either checked out the feature branch, or more likely have just done a `git pull` on `master` and pulled in new config changes. Their current environment is now semi-broken until they do a new `amplify init` to regenerate necessary transient files based on the new files pulled from source control.

I believe this scenario (pulling new files from `master`) is common enough to warrant a doc update. Please correct me though if I've missed something though.

**Comment by armenr** [comment]
I'm still unclear on something - 

Here's what I've done:

1 - Check out new feature branch (from staging to --> staging/myfeatureBranch) in git
2 - Create new env once checked out
3 - Add auth, add storage, add functions
4 - push code + amplify push 

Now, I am ready to merge my feature branch

I go merge my code after reviewing the PR.

Now I am checked back into "staging" branch in Git

How do I merge my BACKEND changes into the "staging" backend? :)
